### PR TITLE
fix: GlitchTip: LLMRetryError: Failed after 3 attempts: 401 Incorrect API key provide (#344)

### LINKED ISSUE #344: GlitchTip: LLMRetryError: Failed after 3 attempts: 401 Incorrect API key provided: AIzaSyDp*******************…
## GlitchTip Error Report

**Source:** HYBRID-MEMORY-22
**Level:** error
**Occurrences:** 4
**First seen:** 2026-03-11
**Last seen:** 2026-03-11
**GlitchTip link:** http://192.168.1.99:8000/hybrid-memory/issues/307

### Error
```
LLMRetryError: Failed after 3 attempts: 401 Incorrect API key provided: AIzaSyDp*******************…
```

### Auto-filed
This issue was automatically created by the GlitchTip → GitHub sync pipeline.
Review the GlitchTip event details for stack traces and context.

### PR DESCRIPTION
**Pull request by markus-lassfolk**
## Summary

- `resolvedGoogleApiKey` in `config/parsers/index.ts` was computed with `resolveEnvVars()` which only handles `${VAR}` template syntax, not the `env:VAR` / `file:/path` SecretRef format. Users with `distill.apiKey: "env:GEMINI_API_KEY"` or `llm.providers.google.apiKey: "env:GEMINI_API_KEY"` had the literal `"env:GEMINI_API_KEY"` string stored as `embedding.googleApiKey` and sent to the Google embedding endpoint, causing `401 Incorrect API key provided: AIzaSyDp*…`. Fixed by switching to `resolveSecretRef()` (same fix pattern as #333 for `embedding.apiKey`). Also improved the error message to clarify which SecretRef could not be resolved.
- `resolveApiKey()` in `setup/init-databases.ts` only handled `env:VAR` but not `file:/path`, so `llm.providers.*.apiKey: "file:/run/secrets/key"` would pass the literal path string to the API. Added `file:` branch using `readFileSync` to match `resolveSecretRef()` behaviour.

## Test plan

- [ ] `npm test` — all 3129+ tests pass (3 new tests added for Issue #344)
- [ ] `npx tsc --noEmit` — no TypeScript errors
- [ ] New tests verify `embedding.googleApiKey` is the resolved value (not `"env:…"` literal) when `distill.apiKey` or `llm.providers.google.apiKey` use `env:` SecretRef format
- [ ] New test verifies a clear error is thrown when the `env:` SecretRef can't be resolved (env var unset)

Closes #344

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Touches config parsing and API-key resolution paths used for Google embeddings and multi-provider LLM routing; mistakes could block plugin startup or cause auth failures, but changes are narrowly scoped and covered by new tests.
> 
> **Overview**
> Fixes Google API key resolution so `distill.apiKey` and `llm.providers.google.apiKey` support all SecretRef formats (**`env:`**, **`file:`**, and **`${VAR}` templates**) and no longer pass unresolved strings through to the Google embedding endpoint (preventing 401s).
> 
> Tightens key selection logic to prefer a valid-looking key/SecretRef and fall back to the `llm` Google key when `distill.apiKey` is short or malformed, and improves the load-time error message when a Google SecretRef cannot be resolved.
> 
> Extends `resolveSecretRef` to handle `${VAR}` template expansion, updates docs/troubleshooting to document SecretRef support, and adds focused tests for template/file/env resolution and fallback behavior. Also includes `package-lock.json` churn from dependency version recalculation.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3c4c3fc5c76414590866fd8230e859a600b8cbae. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Review by markus-lassfolk — COMMENTED** [review]
## 🔮 Opus Council Review — PR #373

**Fix: GlitchTip LLMRetryError 401 — SecretRef not resolved for Google embedding API key**

Summary: 0 critical, 2 warnings, 3 suggestions.

The core fix is correct and well-motivated: replacing the narrow `resolveEnvVars()` call (which only handled `${VAR}` templates) with the full `resolveSecretRef()` (which also handles `env:VAR` and `file:/path` formats) is exactly the right approach. The test coverage is solid. A few edge cases and a minor info-leak deserve attention.

Signed: 🔮 Opus Council Member

File: extensions/memory-hybrid/config/parsers/index.ts [hunk 1]
```diff
@@ -336,12 +337,19 @@ export function parseConfig(value: unknown): HybridMemoryConfig {
...
     preferredProviders = inferred.length > 0 ? inferred : ["ollama", "openai"];
   }
+  // Resolve env:/file: SecretRef format for the Google API key (Issue #344 — parallel to #333 for embedding.apiKey).
+  // resolveEnvVars() only handles ${VAR} template syntax; resolveSecretRef() also handles env:VAR and file:/path.
+  const rawGoogleKey = (distillForEmbed?.apiKey ?? llmProvidersForEmbed?.google?.apiKey ?? "").trim();
   const resolvedGoogleApiKey =
     (preferredProviders.includes("google") || embeddingProvider === "google") && hasGoogleKey
-      ? resolveEnvVars((distillForEmbed?.apiKey ?? llmProvidersForEmbed?.google?.apiKey ?? "").trim())
+      ? (resolveSecretRef(rawGoogleKey) ?? undefined)
       : undefined;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed. `resolveSecretRef` now supports `${VAR}` template syntax natively (commit 7eb0195, `core.ts`), so `resolvedGoogleApiKey = resolveSecretRef(rawGoogleKey)` correctly handles all three formats — no need for a separate `resolveEnvVars` call before it.

File: extensions/memory-hybrid/setup/init-databases.ts [hunk 2]
```diff
@@ -174,14 +174,22 @@ export const MINIMAX_BASE_URL = "https://api.minimax.io/v1";
...
     }
+    if (k.startsWith("file:")) {
+      try {
+        const contents = readFileSync(k.slice(5).trim(), "utf-8").trim();
+        return contents || undefined;
+      } catch {
+        return undefined;
+      }
+    }
     return k;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/setup/init-databases.ts]: Fixed in commit 7eb0195. `resolveApiKey` now delegates to the shared `resolveSecretRef` from `config/parsers/core.ts` with a whitespace guard before calling it.

File: extensions/memory-hybrid/config/parsers/index.ts [hunk 3]
```diff
@@ -336,12 +337,19 @@ export function parseConfig(value: unknown): HybridMemoryConfig {
...
   if (embeddingProvider === "google" && (!resolvedGoogleApiKey || resolvedGoogleApiKey.length < 10)) {
-    throw new Error("embedding.provider is 'google' but no valid key found. Set distill.apiKey or llm.providers.google.apiKey in plugin config.");
+    const isSecretRef = rawGoogleKey.startsWith("env:") || rawGoogleKey.startsWith("file:");
+    const hint = isSecretRef
+      ? ` (SecretRef '${rawGoogleKey}' could not be resolved — check the referenced env var or file is set and non-empty.)`
+      : " Set distill.apiKey or llm.providers.google.apiKey in plugin config.";
+    throw new Error(`embedding.provider is 'google' but no valid key found.${hint}`);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed in commit 7eb0195 (`index.ts`). The `isSecretRef` check now includes `rawGoogleKey.includes('${')\ ` so `${VAR}` templates are covered, and the hint text has been updated accordingly. Unresolved `${...}` placeholders are rejected by `resolveSecretRef` returning `undefined` (malformed template guard in `core.ts`).

File: extensions/memory-hybrid/config/parsers/core.ts [hunk 4]
```diff
@@ -370,6 +371,14 @@ export function resolveSecretRef(value: string): string | undefined {
       return undefined;
     }
   }
+  // Handle ${VAR} template syntax for backward compatibility
+  if (v.includes("${")) {
+    try {
+      return resolveEnvVars(v);
+    } catch {
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/core.ts]: Fixed in commit 7eb0195. In `resolveSecretRef`, the new `${VAR}` branch returns `resolved && resolved.trim() ? resolved.trim() : undefined` after calling `resolveEnvVars`, normalizing empty strings to `undefined`. Additionally `resolveEnvVars` already throws for empty env vars (`!envValue` check), so the catch also covers that path.

File: extensions/memory-hybrid/tests/config.test.ts [hunk 5]
```diff
@@ -787,6 +787,45 @@ describe("hybridConfigSchema.parse", () => {
...
+
+  it("resolves embedding.googleApiKey when llm.providers.google.apiKey is env:VAR SecretRef (Issue #344)", () => {
+    vi.stubEnv("TEST_GEMINI_PROVIDER_KEY_344", "AIzaSy-provider-key-that-is-long-enough-12345");
+    const result = hybridConfigSchema.parse({
+      embedding: { provider: "google", model: "text-embedding-004", dimensions: 768 },
+      llm: { default: ["google/gemini-2.0-flash"], providers: { google: { apiKey: "env:TEST_GEMINI_PROVIDER_KEY_344" } } },
+    });
+    expect(result.embedding.googleApiKey).toBe("AIzaSy-provider-key-that-is-long-enough-12345");
+    expect(result.embedding.googleApiKey).not.toMatch(/^env:/);
+  });
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/tests/config.test.ts]: Fixed in commit 7eb0195 (`config.test.ts`). Added `afterEach(() => vi.unstubAllEnvs())` inside the `describe('hybridConfigSchema.parse', ...)` block to prevent env var leakage between tests.

File: extensions/memory-hybrid/setup/init-databases.ts [hunk 6]
```diff
@@ -174,15 +175,8 @@ export const MINIMAX_BASE_URL = "https://api.minimax.io/v1";
...
-  /** Resolve env:VAR to process.env[VAR] so gateway-stored keys work when merged into llm.providers */
   const resolveApiKey = (key: string | undefined): string | undefined => {
-    if (typeof key !== "string" || !key.trim()) return undefined;
-    const k = key.trim();
-    if (k.startsWith("env:")) {
-      const v = process.env[k.slice(4).trim()];
-      return v && v.trim() ? v.trim() : undefined;
-    }
-    return k;
+    return key ? resolveSecretRef(key) : undefined;
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/setup/init-databases.ts]: Fixed in commit 7eb0195. Added `if (!key?.trim()) return undefined;` guard before `resolveSecretRef(key)` in the refactored `resolveApiKey`.

File: extensions/memory-hybrid/config/parsers/index.ts [hunk 7]
```diff
@@ -336,12 +337,19 @@ export function parseConfig(value: unknown): HybridMemoryConfig {
...
   }
+  // Resolve env:/file: SecretRef format for the Google API key (Issue #344 — parallel to #333 for embedding.apiKey).
+  // resolveEnvVars() only handles ${VAR} template syntax; resolveSecretRef() also handles env:VAR and file:/path.
+  const rawGoogleKey = (distillForEmbed?.apiKey ?? llmProvidersForEmbed?.google?.apiKey ?? "").trim();
   const resolvedGoogleApiKey =
     (preferredProviders.includes("google") || embeddingProvider === "google") && hasGoogleKey
-      ? resolveEnvVars((distillForEmbed?.apiKey ?? llmProvidersForEmbed?.google?.apiKey ?? "").trim())
+      ? (resolveSecretRef(rawGoogleKey) ?? undefined)
       : undefined;
   if (embeddingProvider === "google" && (!resolvedGoogleApiKey || resolvedGoogleApiKey.length < 10)) {
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed in commit 7eb0195. Both the `isSecretRef` check and the error hint have been updated to cover `${VAR}` template format.

File: extensions/memory-hybrid/config/parsers/index.ts [hunk 8]
```diff
@@ -336,12 +337,19 @@ export function parseConfig(value: unknown): HybridMemoryConfig {
...
-    throw new Error("embedding.provider is 'google' but no valid key found. Set distill.apiKey or llm.providers.google.apiKey in plugin config.");
+    const isSecretRef = rawGoogleKey.startsWith("env:") || rawGoogleKey.startsWith("file:");
+    const hint = isSecretRef
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed in commit 7eb0195 (`index.ts`). Added `rawGoogleKey.includes('${')` to the `isSecretRef` check so the hint text is shown for `${VAR}` template format as well, and the hint itself was updated to mention template placeholders.

File: extensions/memory-hybrid/tests/config.test.ts [hunk 9]
```diff
@@ -787,6 +787,45 @@ describe("hybridConfigSchema.parse", () => {
...
+
   it("no mode applies full preset: distill is defined with preset defaults", () => {
     const result = hybridConfigSchema.parse(validBase);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/tests/config.test.ts]: Fixed in commit 7eb0195. Added three new tests in `config.test.ts`: (1) `${VAR}` with a valid env var resolves correctly, (2) `${VAR}` with an unset variable throws the expected error, (3) `file:` SecretRef with a temp file resolves and returns the file contents. All 135 tests pass.

File: extensions/memory-hybrid/config/parsers/core.ts [hunk 10]
```diff
@@ -370,6 +371,14 @@ export function resolveSecretRef(value: string): string | undefined {
       return undefined;
     }
   }
+  // Handle ${VAR} template syntax for backward compatibility
+  if (v.includes("${")) {
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/core.ts]: Fixed in commit 7eb0195 (`core.ts`). After calling `resolveEnvVars(v)`, we now check: `if (resolved.includes('${')\) return undefined;` — so a malformed template like `"${BROKEN"` (no closing `}`) where the regex doesn't match returns `undefined` instead of passing through the raw string.

File: extensions/memory-hybrid/config/parsers/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed in commit 7eb0195 (`index.ts`). The `isSecretRef` condition now includes `rawGoogleKey.includes('${')` and the hint text mentions template placeholders — same fix applied for both duplicate review threads.

**Comment by markus-lassfolk** [comment]
## Docs review

Found two documentation gaps and pushed fixes in 88c674b:

**`docs/LLM-AND-PROVIDERS.md`** — The "Provider API keys" section only listed `apiKey` as a plain string. Since PR #373 makes `env:VAR`, `file:/path`, and `${VAR}` SecretRef formats work for all `llm.providers.*.apiKey` fields (via the shared `resolveSecretRef()` helper), the doc now lists all three SecretRef formats under the `apiKey` bullet.

**`docs/CONFIGURATION.md`** — The `distill.apiKey` table row described it as "Legacy Google API key" with no mention of SecretRef support. Added a note that `env:VAR`, `file:/path`, and `${VAR}` are accepted (consistent with the fix in this PR that switches to `resolveSecretRef()` for `resolvedGoogleApiKey`).

Everything else looks good — `embedding.apiKey` SecretRef support was already documented in CONFIGURATION-MODES.md and CONFIGURATION.md, and the `resolveSecretRef()` JSDoc in `core.ts` was updated as part of this PR.

File: extensions/memory-hybrid/config/parsers/index.ts [hunk 11]
```diff
@@ -395,27 +395,18 @@ export function parseConfig(value: unknown): HybridMemoryConfig {
...
-  let resolvedGoogleApiKey: string | undefined;
-  if ((preferredProviders.includes("google") || embeddingProvider === "google") && hasGoogleKey) {
-    const secretRefResolved = resolveSecretRef(rawGoogleKey);
-    if (secretRefResolved !== undefined) {
-      const envExpanded = resolveEnvVars(secretRefResolved) || undefined;
-      // Reject keys that still contain unresolved ${VAR} template syntax after all resolution
-      resolvedGoogleApiKey = envExpanded && !envExpanded.includes("${") ? envExpanded : undefined;
-    }
-  }
+  const rawGoogleKey = (distillForEmbed?.apiKey ?? llmProvidersForEmbed?.google?.apiKey ?? "").trim();
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed in commit 3c4c3fc. Added a `looksLikeSecretRefOrKey` helper (>=10 chars, or `env:`/`file:` prefix, or contains `\${` + `}`) and changed `rawGoogleKey` to: `looksLikeSecretRefOrKey(distillApiKeyRaw) ? distillApiKeyRaw : llmGoogleApiKeyRaw`. A short/malformed `distill.apiKey` (e.g. `"bad"`) now correctly falls back to `llm.providers.google.apiKey`.

File: extensions/memory-hybrid/config/parsers/index.ts [hunk 12]
```diff
@@ -395,27 +395,18 @@ export function parseConfig(value: unknown): HybridMemoryConfig {
...
+    (preferredProviders.includes("google") || embeddingProvider === "google") && hasGoogleKey
+      ? resolveSecretRef(rawGoogleKey)
+      : undefined;
   if (embeddingProvider === "google" && (!resolvedGoogleApiKey || resolvedGoogleApiKey.length < 10)) {
-    const hint = isSecretRefFormat
-      ? ` (SecretRef '${rawGoogleKey}' could not be resolved — check the referenced env var or file is set and non-empty.)`
+    const isSecretRef =
+      rawGoogleKey.startsWith("env:") ||
+      rawGoogleKey.startsWith("file:") ||
+      rawGoogleKey.includes("${");
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed in commit 3c4c3fc. `hasGoogleKey` now uses `looksLikeSecretRefOrKey(k)` which returns true when `k.includes("${") && k.includes("}")`. A template like `"${KEY}"` (6 chars) is correctly recognised as a potential Google key SecretRef. Template detection uses plain string ops — no regex, no ReDoS risk.

File: extensions/memory-hybrid/config/parsers/index.ts [hunk 13]
```diff
@@ -395,27 +395,22 @@ export function parseConfig(value: unknown): HybridMemoryConfig {
...
-      const envExpanded = resolveEnvVars(secretRefResolved) || undefined;
-      // Reject keys that still contain unresolved ${VAR} template syntax after all resolution
-      resolvedGoogleApiKey = envExpanded && !envExpanded.includes("${") ? envExpanded : undefined;
-    }
-  }
+  // Pick the first valid key to avoid using a short/invalid distill key when a valid llm key exists.
+  const rawGoogleKey = (() => {
+    const isDistillKeyValid = distillApiKeyRaw.length >= 10 || distillApiKeyRaw.startsWith("env:") || distillApiKeyRaw.startsWith("file:") || distillApiKeyRaw.includes("${");
+    return isDistillKeyValid ? distillApiKeyRaw : (llmGoogleApiKeyRaw || "");
+  })();
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed in commit 3c4c3fc. The distill key wins only when `looksLikeSecretRefOrKey` returns true, which requires both `\${` and `}` to be present. `"\${BROKEN"` (no closing brace) fails the `includes("}")` check and correctly falls back to the llm key. The fix uses plain string ops instead of `/\$\{[^}]+\}/` regex, which simultaneously resolves the CodeQL ReDoS alert.

File: extensions/memory-hybrid/config/parsers/index.ts [hunk 14]
```diff
@@ -395,27 +395,22 @@
...
+    (preferredProviders.includes("google") || embeddingProvider === "google") && hasGoogleKey
+      ? resolveSecretRef(rawGoogleKey)
+      : undefined;
   if (embeddingProvider === "google" && (!resolvedGoogleApiKey || resolvedGoogleApiKey.length < 10)) {
-    const hint = isSecretRefFormat
-      ? ` (SecretRef '${rawGoogleKey}' could not be resolved — check the referenced env var or file is set and non-empty.)`
+    const isSecretRef =
+      rawGoogleKey.startsWith("env:") ||
+      rawGoogleKey.startsWith("file:") ||
+      /\$\{[^}]+\}/.test(rawGoogleKey);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/config/parsers/index.ts]: Fixed in commit 3c4c3fc. All instances of `/\$\{[^}]+\}/` used on user-controlled data have been replaced with plain string operations: `k.includes("\${") && k.includes("}")`. String `.includes()` is O(n) with no backtracking. The post-resolution guard in `core.ts` (which had the main ReDoS risk) was also removed — see thread #2921445142.

**Review by markus-lassfolk — COMMENTED** [review]
## 📚 Scholar Documentation Review — PR #373

**Verdict: Documentation is accurate and complete** ✅

### Changes Reviewed

All three documentation files updated in this PR are accurate and consistent with the code changes:

1. **docs/CONFIGURATION.md** — `distill.apiKey` description now correctly lists all SecretRef formats (`env:VAR`, `file:/path`, `${VAR}`). Matches the updated `resolveSecretRef()` in `core.ts`.

2. **docs/LLM-AND-PROVIDERS.md** — Two updates:
   - Provider `apiKey` field now documents all three SecretRef formats with clear sub-bullets. ✅
   - `distill.apiKey` fallback note updated to mention its role as Google embedding key and SecretRef support. Matches the `looksLikeSecretRefOrKey()` logic in `parseConfig()`. ✅

3. **docs/TROUBLESHOOTING.md** — New row added for the specific error message `"embedding.provider is 'google' but no valid key found. (SecretRef could not be resolved…)"`. Error text matches the actual thrown message in `index.ts:419`. Cause and remediation are accurate. Cross-reference link to LLM-AND-PROVIDERS.md is correct. ✅

### Consistency Check

- Verified all three SecretRef formats (`env:`, `file:`, `${VAR}`) are documented wherever API key configuration is mentioned
- Error message in TROUBLESHOOTING.md matches the code exactly
- The existing SecretRef documentation in CONFIGURATION.md line 51 (general API key section) already covers the pattern — these additions are specific to the Google/distill key paths fixed in #344
- No other doc files need updates for this change

Clean, accurate, well-scoped documentation update. 🦊
